[ONNX] Add diagonal symbolic (#64454)

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66149 [ONNX] Update slice process shape to support rank only inference (#65782)
* #66147 [ONNX] Symbolic: dynamic input for OneHot, bool for Einsum (#65940)
* #66146 [ONNX] Modify softplus symbolic to support beta!=1 (#65001)
* #66145 [ONNX] Deprecate fold_if pass (#65697)
* **#66144 [ONNX] Add diagonal symbolic (#64454)**
* #66143 ONNX: Delete or document skipped ORT tests (#64470)
* #66142 [ONNX] Add warning for inplace updates on tensor.shape in tracing mode (#63170)
* #66141 [ONNX] Update sum symbolic to handle dtypes (#64289)
* #66140 [ONNX] Export nn.Module call as ONNX local function (#63589)

* Add logic and tests

* minor edits

* Eliminate expand ops

* Fix flake and editing

* Modified errant message

* Add overrun check

* Add overrun descriptions

* Remove emptyline

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424095](https://our.internmc.facebook.com/intern/diff/D31424095)

shubhambhokare1 committed:
[ONNX] Add diagonal symbolic (#64454)

* Add logic and tests

* minor edits

* Eliminate expand ops

* Fix flake and editing

* Modified errant message

* Add overrun check

* Add overrun descriptions

* Remove emptyline

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Add diagonal symbolic (#64454)"


* Add logic and tests

* minor edits

* Eliminate expand ops

* Fix flake and editing

* Modified errant message

* Add overrun check

* Add overrun descriptions

* Remove emptyline

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/).

BowenBao committed:
Update on "[ONNX] Add diagonal symbolic (#64454)"


* Add logic and tests

* minor edits

* Eliminate expand ops

* Fix flake and editing

* Modified errant message

* Add overrun check

* Add overrun descriptions

* Remove emptyline

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424095](https://our.internmc.facebook.com/intern/diff/D31424095)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Add diagonal symbolic (#64454)"


* Add logic and tests

* minor edits

* Eliminate expand ops

* Fix flake and editing

* Modified errant message

* Add overrun check

* Add overrun descriptions

* Remove emptyline

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424095](https://our.internmc.facebook.com/intern/diff/D31424095)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Add diagonal symbolic (#64454)"


* Add logic and tests

* minor edits

* Eliminate expand ops

* Fix flake and editing

* Modified errant message

* Add overrun check

* Add overrun descriptions

* Remove emptyline

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424095](https://our.internmc.facebook.com/intern/diff/D31424095)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Add diagonal symbolic (#64454)"


* Add logic and tests

* minor edits

* Eliminate expand ops

* Fix flake and editing

* Modified errant message

* Add overrun check

* Add overrun descriptions

* Remove emptyline

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424095](https://our.internmc.facebook.com/intern/diff/D31424095)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Add diagonal symbolic (#64454)"


* Add logic and tests

* minor edits

* Eliminate expand ops

* Fix flake and editing

* Modified errant message

* Add overrun check

* Add overrun descriptions

* Remove emptyline

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D31424095](https://our.internmc.facebook.com/intern/diff/D31424095)

[ghstack-poisoned]
